homeassistant/components/nuimo.py

```diff
@@ -0,0 +1,108 @@
+"""
+components.nuimo
+~~~~~~~~~~~~~~~~~~
+Component that connects to a Bluetooth reachable Nuimo device.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/nuimo/
+"""
+from logging import getLogger
+import threading
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import validate_config
+
+REQUIREMENTS = [
+    'https://github.com/shaftoe/nuimo-raspberrypi-demo/archive/'
+    'pip.zip#nuimo_core==0.0.1',
+    'https://github.com/IanHarvey/bluepy/archive/master.zip#bluepy==1.0.3']
+
+_LOGGER = getLogger(__name__)
+
+DOMAIN = "nuimo"
+EVENT_NUIMO = "nuimo_action"
+
+
+class NuimoDelegate(object):  # pylint: disable=too-few-public-methods
+    """ Translate Nuimo bluetooth messages into HASS events """
+
+    def __init__(self, hass, nuimo):
+        self._nuimo = nuimo
+        self._hass = hass
+
+    # pylint: disable=invalid-name
+    def handleNotification(self, c_handle, data):
+        """ listen for notifications and send them into hass.bus """
+        if int(c_handle) == self._nuimo.characteristicValueHandles['BATTERY']:
+            self._hass.bus.fire(EVENT_NUIMO,
+                                {'action': 'BATTERY', 'value': data[0]})
+        elif int(c_handle) == self._nuimo.characteristicValueHandles['FLY']:
+            self._hass.bus.fire(EVENT_NUIMO, {
+                'action': 'FLY',
+                'value0': data[0],
+                'value1': data[1],
+            })
+        elif int(c_handle) == self._nuimo.characteristicValueHandles['SWIPE']:
+            self._hass.bus.fire(EVENT_NUIMO,
+                                {'action': 'SWIPE', 'value': data[0]})
+        elif int(c_handle) == \
+                self._nuimo.characteristicValueHandles['ROTATION']:
+            value = 'not implemented in Python v3'
+            self._hass.bus.fire(EVENT_NUIMO,
+                                {'action': 'ROTATION', 'value': value})
+        elif int(c_handle) == self._nuimo.characteristicValueHandles['BUTTON']:
+            self._hass.bus.fire(EVENT_NUIMO,
+                                {'action': 'BUTTON', 'value': data[0]})
+
+
+class NuimoManager(threading.Thread):
+    """ Register/unregister Nuimo process """
+
+    def __init__(self, hass, mac):
+        super(NuimoManager, self).__init__()
+        self._hass = hass
+        self._mac = mac
+        self._hass_is_running = True
+
+        from nuimo_core import Nuimo
+        self._nuimo = Nuimo(mac)
+        self._nuimo.set_delegate(NuimoDelegate(self._hass, self._nuimo))
+
+    def run(self):
+        from bluepy.btle import BTLEException
+        try:
```

Comment:
You want to test the connection before the thread starts while in the setup method. That way you can indicate to Home Assistant that setup failed by returning false.